src/Http/Routing/src/Matching/CandidateSet.cs

```diff
@@ -18,19 +18,7 @@ public sealed class CandidateSet
     {
         private const int BitVectorSize = 32;
 
-        // Cannot be readonly because we need to modify it in place.
-        private BitVector32 _validity;
-        private readonly BitArray _largeCapactityValidity;
-
-        // We inline storage for 4 candidates here to avoid allocations in common
-        // cases. There's no real reason why 4 is important, it just seemed like 
-        // a plausible number.
-        private CandidateState _state0;
-        private CandidateState _state1;
-        private CandidateState _state2;
-        private CandidateState _state3;
-
-        private CandidateState[] _additionalCandidates;
```

Comment:
@davidfowl - this boy is two allocations now but much slimmer. Some hot code paths got simpler too.